src/qibo/models/encodings.py

```diff
@@ -184,38 +186,53 @@ def sparse_encoder(data, nqubits: int = None, **kwargs):
     if isinstance(data[0][0], str) and nqubits is None:
         nqubits = len(data[0][0])
 
-    complex_data = bool("complex" in str(type(data[0][1])))
+    _data_test = data[0][1]
+    _data_test = (
+        _data_test.dtype if "array" in str(type(_data_test)) else type(_data_test)
+    )
 
-    # sort data by HW of the bitstrings
-    data_sorted, bitstrings_sorted = _sort_data_sparse(data, nqubits)
+    complex_data = bool("complex" in str(_data_test))
 
+    # sort data by HW of the bitstrings
+    data_sorted, bitstrings_sorted = _sort_data_sparse(data, nqubits, backend)
     # calculate phases
-    _data_sorted = np.abs(data_sorted) if complex_data else data_sorted
-    thetas = _generate_rbs_angles(_data_sorted, architecture="diagonal")
-    phis = np.zeros(len(thetas) + 1)
+    _data_sorted = backend.np.abs(data_sorted) if complex_data else data_sorted
+    thetas = _generate_rbs_angles(
+        _data_sorted, architecture="diagonal", backend=backend
+    )
+    phis = backend.np.zeros(len(thetas) + 1)
     if complex_data:
-        phis[0] = _angle_mod_two_pi(-np.angle(data_sorted[0]))
+        phis[0] = _angle_mod_two_pi(-backend.np.angle(data_sorted[0]))
         for k in range(1, len(phis)):
-            phis[k] = _angle_mod_two_pi(-np.angle(data_sorted[k]) + np.sum(phis[:k]))
+            phis[k] = _angle_mod_two_pi(
+                -backend.np.angle(data_sorted[k]) + backend.np.sum(phis[:k])
+            )
+    phis = backend.cast(phis, dtype=phis[0].dtype)
 
     # marking qubits that have suffered the action of a gate
     touched_qubits = []
-    circuit = comp_basis_encoder(list(bitstrings_sorted[0]), nqubits=nqubits, **kwargs)
+    circuit = comp_basis_encoder(
+        [int(bit) for bit in bitstrings_sorted[0]], nqubits=nqubits, **kwargs
+    )
     for qubit, bit in enumerate(bitstrings_sorted[0]):
         if bit == 1:
-            touched_qubits.append(qubit)
+            touched_qubits.append(int(qubit))
 
     for b_1, b_0, theta, phi in zip(
         bitstrings_sorted[1:], bitstrings_sorted[:-1], thetas, phis
     ):
         hw_0, hw_1 = hamming_weight(b_0), hamming_weight(b_1)
-        distance = hamming_distance("".join(b_1.astype(str)), "".join(b_0.astype(str)))
+        distance = hamming_distance(b_1, b_0)
         difference = b_1 - b_0
 
-        ones, new_ones = np.argsort(b_0)[-hw_0:], np.argsort(b_1)[-hw_1:]
+        ones, new_ones = (
+            list(backend.np.argsort(b_0)[-hw_0:]),
+            list(backend.np.argsort(b_1)[-hw_1:]),
+        )
+        ones, new_ones = {int(elem) for elem in ones}, {int(elem) for elem in new_ones}
```

Comment:
Can't you just do
```suggestion
        ones, new_ones = set(ones), set(new_ones)
```
?

Reply:
This doesn't work again because of GPU backends, without the explicit casting to `int`, the elements are unhashable.